### DISCUSSION TITLE
Adding LOC changed in a commit

### DISCUSSION THREAD

**Discussion by akshaykagarwal** [DLR-SC/gitlab2prov#96] [Ideas]
Is it possible to add the number of lines that were changed(added/removed/modified) per commit activity? This way, one can better track a user's actual activity/contribution instead of just seeing how many commit activity they did. It can also be used to track if an activity was related to a minor change or a major one thus helping in impact analysis as well.

If you can give some pointers on where to look at for updating the code, I can try and update the code :)

**Comment by claasdeboer** [comment]
Yes, this should be possible, will do! I'm going to open an issue from this discussion, this should be part of the next release. The next release will include functionality to extract provenance information from GitHub projects. We still have to figure a few things out behind the scenes before that release is good to go. Will notify you when it goes live.

**Reply by onyame** [reply]
LOC could be an optional [PROV attribute](https://www.w3.org/TR/prov-dm/#generation.attributes) on the activity. @akshaykagarwal Would that help?

**Reply by akshaykagarwal** [reply]
@onyame, yes that would work as well. As long as one can get the LOC per activity, it should be okay. Maybe a flag in the configuration file to select it as part of our output.

**Reply by claasdeboer** [reply]
We can include it by default, it shouldnt have to be a config flag. My idea is to include the amount of LOC changed per commit as PROV attributes for the commit activities as follows:

```python
@dataclass
class GitCommit:
    ...
    insertions: int | None  # number of lines inserted
    deletions: int | None   # number of lines deleted
    lines: int | None       # number of lines changed
    files: int | None       # number of files changed
```

Note that this is only information about how many lines where changed on a commit level and not on a per file level. We can also include information about the number of lines changed per file that were touched in the commit. I think a good place for that could be as attributes on the `ProvGeneration` and `ProvInvalidation` relationships in the [Git: Addition of a File](https://github.com/DLR-SC/gitlab2prov/tree/master/docs#git-addition-of-a-file), [Git: Modification of a File](https://github.com/DLR-SC/gitlab2prov/tree/master/docs#git-addition-of-a-file) and [Git: Deletion of a File](https://github.com/DLR-SC/gitlab2prov/tree/master/docs#git-deletion-of-a-file) models. An alternative place for the number of changes on a file level basis would be as attributes on `FileRevision` entities.

This would include the following information:
```python
@dataclass
class FileRevision:
    ...
    insertions: int | None   # number of lines inserted (for this file)
    deletions: int | None    # number of lines deleted (for this file)
    lines: int | None        # number of lines changed (for this file)
    score: int | None        # percentage of similarity between this and the previous revision
```

**Reply by akshaykagarwal** [reply]
@cdboer I had not even thought of the LOC on a high level. That is actually useful as well along with the LOC per file per commit level.

**Reply by onyame** [reply]
In Gitlab files are not only source code but images or other binary files also. What information should be stored as LOC in that case?

**Reply by akshaykagarwal** [reply]
In that case, it can be just null since LOC for such files wouldn't make sense.

**Reply by claasdeboer** [reply]
> In Gitlab files are not only source code but images or other binary files also. What information should be stored as LOC in that case?

We could think about additional metadata such as file size. But im not sure if that is useful information for data analysis.

**Comment by akshaykagarwal** [comment]
@cdboer, Is there a timeline for the next release that you could share with me (if possible) so I can plan things on my end somewhat based on that?
